docs/cli/ilab-config-structure.md

```diff
@@ -0,0 +1,121 @@
+    data_path: ./taxonomy_data
+    deepspeed_options:
+      cpu_offload_optimizer: true
+      cpu_offload_optimizer_pin_memory: false
+      cpu_offload_optimizer_ratio: 1
+      save_samples: null
+    effective_batch_size: 100
+    is_padding_free: false
+    learning_rate: 2e-6
+    lora:
+      alpha: 32
+      dropout: 0.1
+      quantize_data_type: nf4
+      rank: 2
+      target_modules:
+      - q_proj
+      - k_proj
+      - v_proj
+      - o_proj
+    max_batch_len: 1000
+    max_seq_len: 96
+    mock_data: false
+    mock_data_len: 0
+    model_path: instructlab/granite-7b-lab
+    num_epochs: 1
+    random_seed: 42
+    save_samples: 100
+    warmup_steps: 10
+```
+While useful and clear to the user, this config is hard to maintain, and most users will not care about a large portion of the options.
+
+Keeping some of the key options like `num_epochs`, `deepspeed`, `lora`, and key directories a possible training config could look like:
+
+```yaml
+train:
+  train_args:
+    ckpt_output_dir: checkpoints
+    data_output_dir: train-output
+    data_path: ./taxonomy_data
+    deepspeed_options:
+      cpu_offload_optimizer: true
+      cpu_offload_optimizer_pin_memory: false
+      cpu_offload_optimizer_ratio: 1
+      save_samples: null
+    learning_rate: 2e-6
+    lora:
+      alpha: 32
+      dropout: 0.1
+      quantize_data_type: nf4
+      rank: 2
+      target_modules:
+      - q_proj
+      - k_proj
+      - v_proj
+      - o_proj
+    max_batch_len: 1000
+    model_path: instructlab/granite-7b-lab
+    num_epochs: 1
+    save_samples: 100
+    warmup_steps: 10
+    additional_arguments: ["--is-padding-free=False"...]
+```
+
+`additional_arguments` holds the rest of the training arguments. `ilab` would validate these against an internally maintained list of supported options before passing to the respective library.
+
+The same structure can be applied easily to the serve config.Currently this config looks like: 
+
+```yaml
+serve:
+  backend: ''
+  host_port: 127.0.0.1:8000
+  llama_cpp:
+    gpu_layers: -1
+    llm_family: ''
+    max_ctx_size: 4096
+  model_path: models/merlinite-7b-lab-Q4_K_M.gguf
+  vllm:
+    vllm_args: []
+```
+
+This has the opposite problem as training. The key here is that neither of these options are necessarily wrong, but to have both the verbose structure in the training config juxtaposed against the practically hidden structure of `vllm_args` is not ideal design practice. If we could merge the two approaches to use a common design language that exposes enough key arguments which are commonly edited while also not making the config confusing, this is what we should aim for.
+
+Being very general, this would look something like:
+
+```yaml
+serve:
+  backend: 'vllm'
+  host_port: 127.0.0.1:8000
+  max_ctx_size: 5120
+  gpus: 2
+  llm_family: ''
+  model_path: models/merlinite-7b-lab-Q4_K_M.gguf
+  served_model_name: "merlinite"
+  additional_arguments: ["--block-size=16", "--dtype=fp8"...]
```

Comment:
How would this work if there are multiple commands callable from one config section?  Would we not need the additional_arguments name to indicate where in particular the additional args are being passed through to?

With serving, you could choose to use the same args field to pass values to vllm or llama_cpp or another serving technology.  But If there are two calls to make to vllm or llama_cpp, it becomes less clear.